File: Prolog/models/gameState.pl
:- module(gameState, [ create/6, bird/2, setBird/3, incrementScore/3]).

playingScreenType('playing-screen').
pausedScreenType('paused-screen').
gameOverScreenType('game-over-screen').

create(Bird, PipeGroups, Score, HighestScore, ScreenType, GameState) :-
  isScreenType(ScreenType),
  GameState=gameState(bird(Bird), pipeGroups(PipeGroups), score(Score), highestScore(HighestScore), screenType(ScreenType)).

bird(GameState, Bird) :-
  GameState=gameState(bird(Bird), pipeGroups(_), score(_), highestScore(_), screenType(_)).

pipeGroups(GameState, PipeGroups) :-
  GameState=gameState(bird(_), pipeGroups(PipeGroups), score(_), highestScore(_), screenType(_)).

score(GameState, Score) :-
  GameState=gameState(bird(_), pipeGroups(_), score(Score), highestScore(_), screenType(_)).

highestScore(GameState, HighestScore) :-
  GameState=gameState(bird(_), pipeGroups(_), score(_), highestScore(HighestScore), screenType(_)).

screenType(GameState, ScreenType) :-
  GameState=gameState(bird(_), pipeGroups(_), score(_), highestScore(_), screenType(ScreenType)).

isScreenType(ScreenType) :-
  playingScreenType(ScreenType);
  pausedScreenType(ScreenType);
  gameOverScreenType(ScreenType).

changeScreenType(GameState,NewScreenType,NewGameState):-
  bird(GameState,Bird),
  pipeGroups(GameState,PipeGroups),
  score(GameState,Score),
  highestScore(GameState,HighestScore),
  create(Bird, PipeGroups, Score, HighestScore, NewScreenType, NewGameState).

incrementScore(GameState, Increment, NewGameState) :-
  bird(GameState, Bird),
  pipeGroups(GameState, PipeGroups),
  score(GameState, Score),
  highestScore(GameState, HighestScore),
  screenType(GameState, ScreenType),
  NewScore is Score+Increment,
  NewGameState=gameState(bird(Bird), pipeGroups(PipeGroups), score(NewScore), highestScore(HighestScore), screenType(ScreenType)).

setBird(GameState,NewBird,NewGameState):-
  pipeGroups(GameState,PipeGroups),
  score(GameState,Score),
  highestScore(GameState,HighestScore),
  screenType(GameState,ScreenType),
  create(NewBird, PipeGroups, Score, HighestScore, ScreenType, NewGameState).

File: Haskell/app/Models/GameState.hs
{-# LANGUAGE DataKinds #-}

module Models.GameState where

import Models.Bird (Bird, setVerticalSpeed)
import Models.PipeGroup (PipeGroup)

type PAUSED = "PAUSED"

type PLAYING = "PLAYING"

type GAMEOVER = "GAMEOVER"

data ScreenType = PAUSED | PLAYING | GAMEOVER deriving (Show, Eq)

birdJumpVerticalSpeed :: Float
birdJumpVerticalSpeed = -2

data GameState = GameState
  { bird :: Bird,
    pipeGroups :: [PipeGroup],
    score :: Int,
    highestScore :: Int,
    screenType :: ScreenType
  }

setBird :: GameState -> Bird -> GameState
setBird gameState newBird =
  GameState newBird (pipeGroups gameState) (score gameState) (highestScore gameState) (screenType gameState)

setPipeGroups :: GameState -> [PipeGroup] -> GameState
setPipeGroups gameState newPipeGroups =
  GameState (bird gameState) newPipeGroups (score gameState) (highestScore gameState) (screenType gameState)

incrementScore :: GameState -> Int -> GameState
incrementScore gameState increment =
  GameState (bird gameState) (pipeGroups gameState) (increment + score gameState) (highestScore gameState) (screenType gameState)

setHighestScore :: GameState -> Int -> GameState
setHighestScore gameState newHighestScore =
  GameState (bird gameState) (pipeGroups gameState) (score gameState) newHighestScore (screenType gameState)

setScreenType :: GameState -> ScreenType -> GameState
setScreenType gameState = GameState (bird gameState) (pipeGroups gameState) (score gameState) (highestScore gameState)

jumpBird :: GameState -> Bird -> GameState
jumpBird state bird =
  setBird state (setVerticalSpeed bird birdJumpVerticalSpeed)
